br/pkg/utils/retry.go

```diff
@@ -32,6 +36,130 @@ var retryableServerError = []string{
 	"end of file before message length reached",
 }
 
+type ErrorResult struct {
+	Strategy ErrorStrategy
+	Reason   string
+}
+
+type ErrorStrategy int
+
+const (
+	// This type can be retry but consume the backoffer attempts.
+	Retry ErrorStrategy = iota
+	// This type means un-recover error and the whole progress should exits
+	// for example:
+	// 1. permission not valid.
+	// 2. data has not found.
+	// 3. retry too many times
+	GiveUp
+)
+
+type ErrorContext struct {
+	mu sync.Mutex
+	// encounter times for one context on a store
+	// we may use this value to determine the retry policy
+	encounterTimes map[uint64]int
+	// unknown error retry limitation.
+	// encouter many times error makes Retry to GiveUp.
+	encounterTimesLimitation int
+	// whether in backup or restore
+	scenario string
+}
+
+func NewErrorContext(scenario string, limitation int) *ErrorContext {
+	return &ErrorContext{
+		scenario:                 scenario,
+		encounterTimes:           make(map[uint64]int),
+		encounterTimesLimitation: limitation,
+	}
+}
+
+func NewDefaultContext() *ErrorContext {
+	return &ErrorContext{
+		encounterTimes:           make(map[uint64]int),
+		encounterTimesLimitation: 1,
+	}
+}
+
+func (ec *ErrorContext) HandleError(err *backuppb.Error, uuid uint64, canIgnore bool) ErrorResult {
+	if len(err.Msg) != 0 {
+		return ec.HandleErrorMsg(err.Msg, uuid)
+	}
+	return ec.HandleErrorPb(err, uuid, canIgnore)
+}
+
+func (ec *ErrorContext) HandleErrorMsg(msg string, uuid uint64) ErrorResult {
+	// UNSAFE! TODO: use meaningful error code instead of unstructured message to find failed to write error.
+	logger := log.L().With(zap.String("scenario", ec.scenario))
+	if messageIsNotFoundStorageError(msg) {
+		reason := fmt.Sprintf("File or directory not found on TiKV Node (store id: %v). "+
+			"work around:please ensure br and tikv nodes share a same storage and the user of br and tikv has same uid.",
+			uuid)
+		return ErrorResult{GiveUp, reason}
+	}
+	if messageIsPermissionDeniedStorageError(msg) {
+		reason := fmt.Sprintf("I/O permission denied error occurs on TiKV Node(store id: %v). "+
+			"work around:please ensure tikv has permission to read from & write to the storage.",
+			uuid)
+		return ErrorResult{GiveUp, reason}
+	}
+
+	if MessageIsRetryableStorageError(msg) {
+		logger.Warn("occur storage error", zap.String("error", msg))
+		return ErrorResult{Retry, "retrable error"}
+	}
+	// retry enough on same store
+	mu.Lock()
+	defer mu.Unlock()
+	ec.encounterTimes[uuid]++
+	if ec.encounterTimes[uuid] < ec.encounterTimesLimitation {
+		return ErrorResult{Retry, "unknown error, retry it for few times"}
+	}
+	return ErrorResult{GiveUp, "unknown error and retry too many times, give up"}
+}
+
+func (ec *ErrorContext) HandleErrorPb(e *backuppb.Error, uuid uint64, canIgnore bool) ErrorResult {
```

Comment:
`canIgnore` looks so magical here, I guess it should be the requirement of the coarse-grained backup. I will prefer make another function to handle the scenario.

Reply:
`pushBackup` need it, but `fine-grained` doesn't. Do you think make it a field of `errContext` is better?

Reply:
I think we can make two functions: `HandleErrAndIgnoreStoreErrors` and `HandleErr` or something. 🤔